pkg/storage/replica.go

```diff
@@ -1936,10 +1951,10 @@ func (r *Replica) tryAddWriteCmd(
 			if tryAbandon() {
 				log.Warningf(ctx, "context cancellation after %0.1fs of attempting command %s",
 					timeutil.Since(startTime).Seconds(), ba)
-				// Set endCmds to nil because they will be invoked in processRaftCommand.
-				endCmds = nil
 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError(ctx.Err().Error())), proposalNoRetry
 			}
+			// Set endCmds to nil because they will be invoked in processRaftCommand.
```

Comment:
The code here was correct before. We cannot end the commands by exiting this method early with `endCmds` still non-nil. This would incorrectly unblock the command queue, possibly allowing overlapping commands to execute concurrently.

What needs to be changed is code in `tryAbandon`. The whole point of this `tryAbandon` mechanism is to avoid use of a context which might have already had its trace span finalized when processing the raft function. I think `tryAbandon` needs to be modified to ensure that we continue to keep the `EvalResult` in the `Replica.mu.replicas` map (with `endCmds` left set), but just change something to indicate that the context can no longer be used when `processRaftCommand` is executed.